src/Vgrep/Ansi/Parser.hs

```diff
@@ -0,0 +1,145 @@
+
+Non-CSI sequences are not parsed, but included in the output:
+
+>>> parseAnsi "\ESC]710;font\007foo\ESC[31mbar"
+Cat 17 [Text 14 "\ESC]710;font\afoo",Format 3 (Attr {attrStyle = KeepCurrent, attrForeColor = SetTo (ISOColor 1), attrBackColor = KeepCurrent}) (Text 3 "bar")]
+
+-}
+parseAnsi :: Text -> Formatted Attr
+parseAnsi = either error id . parseOnly ansiFormatted
+-- The use of 'error' ↑ is safe: 'ansiFormatted' does not fail.
+
+
+-- | Parser for ANSI formatted text. Recognized escape sequences are the SGR
+-- (Select Graphic Rendition) sequences (@\ESC[…m@) supported by 'Attr'.
+-- Unsupported SGR sequences and other CSI escape sequences (@\ESC[…@) are
+-- ignored. Other (non-CSI) escape sequences are not parsed, and included in the
+-- output.
+--
+-- This parser does not fail, it will rather consume and return the remaining
+-- input as unformatted text.
+ansiFormatted :: Parser (Formatted Attr)
+ansiFormatted = go mempty
+  where
+    go attr = endOfInput *> pure mempty
+          <|> formattedText attr
+    formattedText attr = do
+        acs <- many attrChange
+        let attr' = foldr ($) attr acs
+        t <- rawText
+        rest <- go attr'
+        pure (format attr' (bare t) <> rest)
+    rawText = atLeastOneTill (== '\ESC') <|> endOfInput *> pure ""
+    atLeastOneTill = liftA2 T.cons anyChar . takeTill
+
+
+-- | Parser for ANSI CSI escape sequences. Recognized escape sequences are the
+-- SGR (Select Graphic Rendition) sequences (@\ESC[…m@) supported by 'Attr'.
+-- Unsupported SGR sequences and other CSI escape sequences (@\ESC[…@) are
+-- ignored by returning 'id'.
+--
+-- This parser fails when encountering any other (non-CSI) escape sequence.
+attrChange :: Parser (Attr -> Attr)
+attrChange = fmap csiToAttrChange csi
+
+csiEscape :: Parser Text
+csiEscape = "\ESC["
+
+csi :: Parser Csi
+csi = csiEscape >> liftA2 Csi (decimal `sepBy` char ';') anyChar
+
+data Csi = Csi [Int] Char
+
+csiToAttrChange :: Csi -> Attr -> Attr
+csiToAttrChange = \case
+    Csi [] 'm' -> const mempty
+    Csi is 'm' -> foldMap attrChangeFromCode is
+    _otherwise -> id
+
+attrChangeFromCode :: Int -> Attr -> Attr
+attrChangeFromCode = \case
+    0  -> const mempty
+    1  -> withStyle Vty.bold
+    3  -> withStyle Vty.standout
+    4  -> withStyle Vty.underline
+    5  -> withStyle Vty.blink
+    6  -> withStyle Vty.blink
+    7  -> withStyle Vty.reverseVideo
+    21 -> withoutStyle Vty.bold
+    22 -> withoutStyle Vty.bold
+    23 -> withoutStyle Vty.standout
+    24 -> withoutStyle Vty.underline
+    25 -> withoutStyle Vty.blink
+    27 -> withoutStyle Vty.reverseVideo
+    i | i >= 30  && i <= 37   -> withForeColor (rawColor (i - 30))
+      | i >= 40  && i <= 47   -> withBackColor (rawColor (i - 40))
+      | i >= 90  && i <= 97   -> withForeColor (rawBrightColor (i - 90))
+      | i >= 100 && i <= 107  -> withBackColor (rawBrightColor (i - 100))
```

Comment:
Yes, I like to add a bit of visual separation between the guard and the implementation.